File: internal/dto/job.go
package dto

import (
	"Postulator/internal/domain/job"
	"time"
)

type Job struct {
	ID           int64  `json:"id"`
	Name         string `json:"name"`
	SiteID       int64  `json:"siteId"`
	CategoryID   int64  `json:"categoryId"`
	PromptID     int64  `json:"promptId"`
	AIProviderID int64  `json:"aiProviderId"`
	AIModel      string `json:"aiModel"`

	RequiresValidation bool    `json:"requiresValidation"`
	ScheduleType       string  `json:"scheduleType"`
	ScheduleTime       *string `json:"scheduleTime,omitempty"`
	ScheduleDay        *int    `json:"scheduleDay,omitempty"`
	JitterEnabled      bool    `json:"jitterEnabled"`
	JitterMinutes      int     `json:"jitterMinutes"`
	Status             string  `json:"status"`
	LastRunAt          *string `json:"lastRunAt,omitempty"`
	NextRunAt          *string `json:"nextRunAt,omitempty"`
	CreatedAt          string  `json:"createdAt"`
	UpdatedAt          string  `json:"updatedAt"`
}

type Execution struct {
	ID               int64   `json:"id"`
	JobID            int64   `json:"jobId"`
	TopicID          int64   `json:"topicId"`
	GeneratedTitle   *string `json:"generatedTitle,omitempty"`
	GeneratedContent *string `json:"generatedContent,omitempty"`
	Status           string  `json:"status"`
	ErrorMessage     *string `json:"errorMessage,omitempty"`
	ArticleID        *int64  `json:"articleId,omitempty"`
	StartedAt        string  `json:"startedAt"`
	GeneratedAt      *string `json:"generatedAt,omitempty"`
	ValidatedAt      *string `json:"validatedAt,omitempty"`
	PublishedAt      *string `json:"publishedAt,omitempty"`
}

func FromJob(e *job.Job) *Job {
	if e == nil {
		return nil
	}
	toTimeStr := func(t *time.Time) *string {
		if t == nil {
			return nil
		}
		v := t.UTC().Format(timeLayout)
		return &v
	}
	var scheduleTime *string
	if e.ScheduleTime != nil {
		v := e.ScheduleTime.UTC().Format("15:04:05")
		scheduleTime = &v
	}
	return &Job{
		ID:                 e.ID,
		Name:               e.Name,
		SiteID:             e.SiteID,
		CategoryID:         e.CategoryID,
		PromptID:           e.PromptID,
		AIProviderID:       e.AIProviderID,
		AIModel:            e.AIModel,
		RequiresValidation: e.RequiresValidation,
		ScheduleType:       string(e.ScheduleType),
		ScheduleTime:       scheduleTime,
		ScheduleDay:        e.ScheduleDay,
		JitterEnabled:      e.JitterEnabled,
		JitterMinutes:      e.JitterMinutes,
		Status:             string(e.Status),
		LastRunAt:          toTimeStr(e.LastRunAt),
		NextRunAt:          toTimeStr(e.NextRunAt),
		CreatedAt:          e.CreatedAt.UTC().Format(timeLayout),
		UpdatedAt:          e.UpdatedAt.UTC().Format(timeLayout),
	}
}

func FromJobs(items []*job.Job) []*Job {
	out := make([]*Job, 0, len(items))
	for _, it := range items {
		out = append(out, FromJob(it))
	}
	return out
}

func FromExecution(e *job.Execution) *Execution {
	if e == nil {
		return nil
	}
	toTimeStr := func(t *time.Time) *string {
		if t == nil {
			return nil
		}
		v := t.UTC().Format(timeLayout)
		return &v
	}
	return &Execution{
		ID:               e.ID,
		JobID:            e.JobID,
		TopicID:          e.TopicID,
		GeneratedTitle:   e.GeneratedTitle,
		GeneratedContent: e.GeneratedContent,
		Status:           string(e.Status),
		ErrorMessage:     e.ErrorMessage,
		ArticleID:        e.ArticleID,
		StartedAt:        e.StartedAt.UTC().Format(timeLayout),
		GeneratedAt:      toTimeStr(e.GeneratedAt),
		ValidatedAt:      toTimeStr(e.ValidatedAt),
		PublishedAt:      toTimeStr(e.PublishedAt),
	}
}

func FromExecutions(items []*job.Execution) []*Execution {
	out := make([]*Execution, 0, len(items))
	for _, it := range items {
		out = append(out, FromExecution(it))
	}
	return out
}

File: frontend/src/services/job.ts
import {
  CreateJob,
  DeleteJob,
  ExecuteJobManually,
  GetJob,
  GetPendingValidations,
  ListJobs,
  PauseJob,
  RestoreScheduler,
  ResumeJob,
  UpdateJob,
  ValidateExecution,
} from "@/wailsjs/wailsjs/go/app/App";
import { dto } from "@/wailsjs/wailsjs/go/models";
import { unwrapMany, unwrapOne, unwrapString } from "./utils";

import type { ScheduleType, JobStatus, ExecutionStatus } from "@/constants/jobs";

export interface Job {
  id: number;
  name: string;
  siteId: number;
  categoryId: number;
  promptId: number;
  aiProviderId: number;
  aiModel: string;
  requiresValidation: boolean;
  scheduleType: ScheduleType;
  intervalValue?: number;
  intervalUnit?: string;
  scheduleHour?: number;
  scheduleMinute?: number;
  weekdays?: number[];
  monthdays?: number[];
  jitterEnabled: boolean;
  jitterMinutes: number;
  status: JobStatus;
  lastRunAt?: string;
  nextRunAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Execution {
  id: number;
  jobId: number;
  topicId: number;
  generatedTitle?: string;
  generatedContent?: string;
  status: ExecutionStatus;
  errorMessage?: string;
  articleId?: number;
  startedAt: string;
  generatedAt?: string;
  validatedAt?: string;
  publishedAt?: string;
}

function mapJob(x: dto.Job): Job {
  return {
    id: x.id,
    name: x.name,
    siteId: x.siteId,
    categoryId: x.categoryId,
    promptId: x.promptId,
    aiProviderId: x.aiProviderId,
    aiModel: x.aiModel,
    requiresValidation: x.requiresValidation,
    scheduleType: x.scheduleType as ScheduleType,
    intervalValue: x.intervalValue,
    intervalUnit: x.intervalUnit,
    scheduleHour: x.scheduleHour,
    scheduleMinute: x.scheduleMinute,
    weekdays: x.weekdays,
    monthdays: x.monthdays,
    jitterEnabled: x.jitterEnabled,
    jitterMinutes: x.jitterMinutes,
    status: x.status as JobStatus,
    lastRunAt: x.lastRunAt,
    nextRunAt: x.nextRunAt,
    createdAt: x.createdAt,
    updatedAt: x.updatedAt,
  };
}

function mapExecution(x: dto.Execution): Execution {
  return {
    id: x.id,
    jobId: x.jobId,
    topicId: x.topicId,
    generatedTitle: x.generatedTitle,
    generatedContent: x.generatedContent,
    status: x.status as ExecutionStatus,
    errorMessage: x.errorMessage,
    articleId: x.articleId,
    startedAt: x.startedAt,
    generatedAt: x.generatedAt,
    validatedAt: x.validatedAt,
    publishedAt: x.publishedAt,
  };
}

export async function listJobs(): Promise<Job[]> {
  const res = await ListJobs();
  return unwrapMany<dto.Job>(res).map(mapJob);
}

export async function getJob(id: number): Promise<Job> {
  const res = await GetJob(id);
  return mapJob(unwrapOne<dto.Job>(res));
}

export async function createJob(input: Omit<Job, "id" | "createdAt" | "updatedAt" | "lastRunAt" | "nextRunAt" | "status"> & { status?: string }): Promise<string> {
  const payload = new dto.Job({ ...input });
  const res = await CreateJob(payload);
  return unwrapString(res);
}

export async function updateJob(input: Omit<Job, "createdAt" | "updatedAt">): Promise<string> {
  const payload = new dto.Job({ ...input });
  const res = await UpdateJob(payload);
  return unwrapString(res);
}

export async function deleteJob(id: number): Promise<string> {
  const res = await DeleteJob(id);
  return unwrapString(res);
}

export async function executeJobManually(id: number): Promise<string> {
  const res = await ExecuteJobManually(id);
  return unwrapString(res);
}

export async function pauseJob(id: number): Promise<string> {
  const res = await PauseJob(id);
  return unwrapString(res);
}

export async function resumeJob(id: number): Promise<string> {
  const res = await ResumeJob(id);
  return unwrapString(res);
}

export async function restoreScheduler(): Promise<string> {
  const res = await RestoreScheduler();
  return unwrapString(res);
}

export async function getPendingValidations(): Promise<Execution[]> {
  const res = await GetPendingValidations();
  return unwrapMany<dto.Execution>(res).map(mapExecution);
}

export async function validateExecution(executionId: number, approved: boolean): Promise<string> {
  const res = await ValidateExecution(executionId, approved);
  return unwrapString(res);
}
